python/activator/middleware_interface.py

```diff
@@ -176,42 +184,122 @@ def prep_butler(self, visit: Visit) -> None:
+            Center of the region to load the skyamp tract/patches for.
+        detector : `lsst.afw.cameraGeom.Detector`
+            Detector we are loading data for.
+        wcs : `lsst.afw.geom.SkyWcs`
+            Rough WCS for the upcoming visit, to help finding patches.
+        """
+        export.saveDatasets(self.central_butler.registry.queryDatasets("skyMap",
+                                                                       collections="skymaps",
+                                                                       findFirst=True))
+        # Getting only one tract should be safe: we're getting the
+        # tract closest to this detector, so we should be well within
+        # the tract bbox.
+        skymap = self.central_butler.get("skyMap")
+        tract = skymap.findTract(center)
+        points = [center]
+        for corner in detector.getCorners(lsst.afw.cameraGeom.PIXELS):
+            points.append(wcs.pixelToSky(corner))
+        patches = tract.findPatchList(points)
+        template_where = f"patch in ({','.join(str(p.sequential_index) for p in patches)}) and tract={tract.tract_id}"
+        # TODO: do we need to have the coadd name used in the pipeline
+        # specified as a class kwarg, so that we only load one here?
+        # TODO: alternately, we need to extract it from the pipeline?
+        # TODO: alternately, can we just assume that there is exactly
+        # one coadd in the central butler?
+        export.saveDatasets(self.central_butler.registry.queryDatasets("*Coadd",
+                                                                       collections="templates",
+                                                                       where=template_where))
+
+    def _export_calibs(self, export, detector_id, filter):
+        """Export the calibs for this visit from the central butler.
+
+        Parameters
+        ----------
+        export : `Iterator[RepoExportContext]`
+            Export context manager.
+        detector_id : `int`
+            Identifier of the detector to load calibs for.
+        filter : `str`
+            Physical filter name of the upcoming visit.
+        """
+        # TODO: we can't filter by validity range because it's not
+        # supported in queryDatasets yet.
+        calib_where = f"detector={detector_id} and physical_filter='{filter}'"
+        # import ipdb; ipdb.set_trace()
+        export.saveDatasets(self.central_butler.registry.queryDatasets(
+            ...,
+            collections=self.instrument.makeCalibrationCollectionName(),
+            where=calib_where))
+        target_types = {CollectionType.CALIBRATION}
+        for collection in self.central_butler.registry.queryCollections(...,
+                                                                        collectionTypes=target_types):
+            export.saveCollection(collection)
+
+        # TODO: Should probably also get the appropriate chains here, too?
```

Comment:
I guess if you provide all run collections to the pipeline executor, the chains don't matter.

Reply:
Providing runs to the pipeline is going to be impractical in the long run, because they depend on where exactly the data came from (and this goes for test data as well). While they're hard to export because you need exact definitions, chains make it *much* easier to program the pipeline parts against a specific interface.

Reply:
We absolutely want to program assuming a default set of chains. That's what they're for. I think I've got it now anyway.